### ISSUE TITLE
missing second argument record in struct

### ISSUE THREAD

**Issue by honggyukim** [namhyung/uftrace#1281]
Here is an example.
```cpp
$ cat iters.cc
#include <vector>

__attribute__((noinline))
int compare_iters(std::vector<int>::iterator beg,
                  std::vector<int>::iterator end)
{
        return beg == end;
}

int main()
{
    std::vector<int> v(3, 1);
    return compare_iters(v.begin(), v.end());
}
```
We can compile and run this program as follows.
```
$ g++ -pg -g -O2 iters.cc

$ uftrace -a a.out
# DURATION     TID     FUNCTION
            [1279330] | main() {
   1.958 us [1279330] |   operator new(12) = 0x264ab760;
   0.166 us [1279330] |   compare_iters(__normal_iterator{...}) = 0;
   1.834 us [1279330] |   operator delete(0x264ab760);
   5.959 us [1279330] | } = 0; /* main */
```
The result shows that `compare_iters` shows only a single argument, but it actually has two arguments as written in the source code above.

The `a.out.dbg` shows it has two arguments.
```
$ uftrace record -a --no-event a.out

$ cat uftrace.data/a.out.dbg
      ...
F: 8f0 compare_iters
L: 5 iters.cc
A: @arg1/t8:__normal_iterator%x0,arg2/t8:__normal_iterator%x1
R: @retval
```
The dump output clearly shows that uftrace recorded the first argument only.
```
$ uftrace dump -v
uftrace: running uftrace v0.10-8-g4bfe0 ( aarch64 dwarf python luajit tui perf sched dynamic )
    ...
uftrace file header: features      = 0x67b (PLTHOOK | TASK_SESSION | ARGUMENT | RETVAL | SYM_REL_ADDR | MAX_STACK | AUTO_ARGS | DEBUG_INFO)
    ...
1464440.253534694  1279809: [entry] compare_iters(400910) depth: 1
 <0000000000000040>: e6 51 1b 91 e6 33 05 00  6c 00 10 09 40 00 00 00
1464440.253534694  1279809: [args ] length = 8
  args[0] struct __normal_iterator:
        1c f2 41 2c 00 00 00 00
 <0000000000000050>: 1c f2 41 2c 00 00 00 00
1464440.253534819  1279809: [exit ] compare_iters(400910) depth: 1
 <0000000000000058>: 63 52 1b 91 e6 33 05 00  6d 00 10 09 40 00 00 00
1464440.253534819  1279809: [retval] length = 8
  retval d64: 0x0000000000000000
 <0000000000000068>: 00 00 00 00 00 00 00 00
```
We need to figure out why it didn't record the second argument.

**Comment by honggyukim** [comment]
It's tested on aarch64, but need to confirm on x86_64 as well.

**Comment by honggyukim** [comment]
x86_64 has the same problem and its debug info is as follows.
```
F: 7e0 compare_iters
L: 5 iters.cc
A: @arg1/t8:__normal_iterator%rdi,arg2/t8:__normal_iterator%rsi
R: @retval
```

**Comment by namhyung** [comment]
It seems it needs to update arg->reg_idx for struct passed by registers, otherwise the second argspec just overwrote the first one as if it founds a same spec.  The below patch seems to fix it for me

```diff
diff --git a/utils/argspec.c b/utils/argspec.c
index fda74c3a..29f9560c 100644
--- a/utils/argspec.c
+++ b/utils/argspec.c
@@ -170,8 +170,10 @@ struct uftrace_arg_spec * parse_argspec(char *str,
                                        *next = '\0';
 
                                reg = arch_register_number(setting->arch, ++suffix);
-                               if (reg >= 0)
+                               if (reg >= 0) {
                                        arg->struct_regs[arg->struct_reg_cnt++] = reg;
+                                       arg->reg_idx = reg;
+                               }
 
                                suffix = next;
                        }
```

**Comment by honggyukim** [comment]
@namhyung Thanks. I see that it fixes the problem.
```
$ uftrace -a a.out
# DURATION     TID     FUNCTION
            [ 22790] | main() {
   2.234 us [ 22790] |   operator new(12) = 0x1aecb80;
 113.048 us [ 22790] |   compare_iters(__normal_iterator{...}, __normal_iterator{...}) = 0;
   1.717 us [ 22790] |   operator delete(0x1aecb80);
 120.732 us [ 22790] | } = 0; /* main */
```

**Comment by honggyukim** [comment]
@namhyung Thanks for doing this fix! But I see 1398d5e is a breaking change. If there is previously recorded data then it can't be replayed normally.

We can record and replay with the previous version.
```
$ uftrace record -a a.out

$ uftrace replay
# DURATION     TID     FUNCTION
            [ 43462] | main() {
   1.935 us [ 43462] |   operator new(12) = 0xed3e90;
   0.324 us [ 43462] |   compare_iters(__normal_iterator{...}) = 0;
   1.714 us [ 43462] |   operator delete(0xed3e90);
   7.418 us [ 43462] | } = 0; /* main */
```
But after updating uftrace with the fix 1398d5e, it fails to show replay.
```
$ uftrace replay
WARN: invalid rstack read
# DURATION     TID     FUNCTION
            [ 43462] | main() {
   1.935 us [ 43462] |   operator new(12) = 0xed3e90;
            [ 43462] |   compare_iters(__normal_iterator{...}, __normal_iterator{...}) {
            [ 43462] |     /* linux:task-exit */

uftrace stopped tracing with remaining functions
================================================
task: 43462
[1] compare_iters
[0] main
```
I know it's unavoidable but I just leave this comment for the record.
